Expose the mapper to derivers, moving some code out of main.

paurkedal opened the pull request:
Ad issue #33, maybe some final adjustments?

whitequark commented on src/ppx_deriving.ml:
```diff
@@ -2,6 +2,7 @@ open Longident
 open Location
 open Asttypes
 open Parsetree
+open Ast_mapper
```
You don't have to open the Ast_mapper, all references are qualified

On the other hand you can even remove Ast_helper below

paurkedal replied:
They are both used multiple places in `ppx_deriving`, esp. I would prefer to keep Ast_mapper due to the record field uses.  OTOH, I could remove all open statements from `ppx_deriving_main`.

whitequark replied:
Oh. Maybe just add Ast_mapper.{} around that? I would really prefer not to have get_cookie and such in global scope, it's dirty enough anyway.

whitequark commented on src/ppx_deriving.ml:
```diff
@@ -304,3 +305,181 @@ let extract_typename_of_type_group deriver ~allow_shadowing type_list =
+  List.iter dynlink plugins;
+  let loaded  = loaded @ plugins in
+  Ast_mapper.set_cookie "ppx_deriving"
+    (Exp.tuple (List.map (fun file -> Exp.constant (Const_string (file, None))) loaded))
```
All Dynlink-related stuff should live in Ppx_deriving_main. Ppx_deriving and the library shouldn't even depend on Dynlink

paurkedal replied:
Are you sure?  There is a match on @@@findlib.ppxopt in the `structure` local function which calls `add_plugins` and which calls `dynlink`.  I guess it's there to support the `#ppx_deriving` directives in the toplevel.

whitequark replied:
Oh, yeah, that should also go into `_main`. `Ppx_deriving` itself should really be free of any ocaml/ocamlc-related invocation issues, that would make reusing it much cleaner.

whitequark replied:
Basically, `Ppx_deriving` is an AST mapper, `Ppx_deriving_main` is an AST mapper *specifically for the default OCaml toolchain*.

paurkedal commented:
I've qualified all Ast_mapper reference and moved dynlink stuff back to main.  In doing the latter, I changed recursion of structure and signature so that they use the mapper.  That allowed me to hook in the ppxopt from main.  I'm not sure how to test the toplevel support though.  I guessed `#ppx "deriving"` but I'm not sure how ppxopt is passed.

whitequark commented:
Nope. Toplevel is tested in the exact same way as non-toplevel: with findlib. You do `#require "ppx_deriving.std";` for example.

paurkedal commented:
That explains why `#ppx` is not documented. Digging around a bit, I see that `topfind.ml.in` injects `@@@findlib.ppxopt` to communicate with the syntax extension.

whitequark commented:
I verified it works on toplevel. Merging, thanks!

whitequark commented:
It is an overlook that `#ppx` is not documented. It is equivalent to a documented `ocamlc -ppx` option.

paurkedal commented:
Right, I did test that, as well. And I also have uncomitted code which uses the mapper.

Thanks!
